frame/two-phase-election-provider/src/benchmarking.rs

```diff
@@ -0,0 +1,358 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2020 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Two phase election pallet benchmarking.
+
+use super::*;
+use crate::Module as TwoPhase;
+
+pub use frame_benchmarking::{account, benchmarks, whitelist_account, whitelisted_caller};
+use frame_support::{assert_ok, traits::OnInitialize};
+use frame_system::RawOrigin;
+use rand::{prelude::SliceRandom, rngs::SmallRng, SeedableRng};
+use sp_election_providers::Assignment;
+use sp_npos_elections::ExtendedBalance;
+use sp_runtime::InnerOf;
+use sp_arithmetic::traits::One;
+use sp_std::convert::TryInto;
+
+const SEED: u32 = 0;
+
+/// Creates a **valid** solution with exactly the given size.
+///
+/// The snapshot is also created internally.
+fn solution_with_size<T: Config>(
+	size: SolutionSize,
+	active_voters_count: u32,
+	winners_count: u32,
+) -> RawSolution<CompactOf<T>>
+where
+	ExtendedBalance: From<InnerOf<CompactAccuracyOf<T>>>,
+	ExtendedBalance: From<InnerOf<OnChainAccuracyOf<T>>>,
+	<InnerOf<CompactAccuracyOf<T>> as sp_std::convert::TryFrom<usize>>::Error: sp_std::fmt::Debug,
+{
+	assert!(size.targets >= winners_count, "must have enough targets");
+	assert!(
+		size.targets >= (<CompactOf<T>>::LIMIT * 2) as u32,
+		"must have enough targets for unique votes."
+	);
+	assert!(
+		size.voters >= active_voters_count,
+		"must have enough voters"
+	);
+	assert!(
+		(<CompactOf<T>>::LIMIT as u32) < winners_count,
+		"must have enough winners to give them votes."
+	);
```

Comment:
in the future, returning a result i think gives a better output when running benchmarks and hitting errors

so these would generally be `ensure!` for me

Reply:
Better output in which sense? These are the assumption that if broken the benchmarks should not run. Will returning error also fit that?

Reply:
I think it might tell you which benchmark failed rather than just a panic message, anyway not a big deal.